src/main/java/bio/terra/stairway/Queue.java

```diff
@@ -0,0 +1,237 @@
+            publisher.shutdown();
+        }
+    }
+
+    public void dispatchMessages(int numOfMessages, QueueProcessFunction<String, Stairway, Boolean> processFunction)
+            throws InterruptedException {
+
+        PullRequest pullRequest =
+                PullRequest.newBuilder()
+                        .setMaxMessages(numOfMessages)
+                        .setReturnImmediately(false)
+                        .setSubscription(subscriptionName)
+                        .build();
+
+        // use pullCallable().futureCall to asynchronously perform this operation
+        // The call can complete without returning any messages.
+        PullResponse pullResponse = subscriber.pullCallable().call(pullRequest);
+        if (pullResponse.getReceivedMessagesList().size() == 0) {
+            return;
+        }
+
+        List<String> ackIds = new ArrayList<>();
+        for (ReceivedMessage message : pullResponse.getReceivedMessagesList()) {
+            String smess = message.getMessage().getData().toStringUtf8();
+            logger.info("Received message: " + smess);
+
+            Boolean processSucceeded = processFunction.apply(smess, stairway);
+            if (processSucceeded) {
+                ackIds.add(message.getAckId());
+            }
+        }
+
+        if (ackIds.size() > 0) {
+            // acknowledge received messages
+            AcknowledgeRequest acknowledgeRequest =
+                    AcknowledgeRequest.newBuilder()
+                            .setSubscription(subscriptionName)
+                            .addAllAckIds(ackIds)
+                            .build();
+            // use acknowledgeCallable().futureCall to asynchronously perform this operation
+            subscriber.acknowledgeCallable().call(acknowledgeRequest);
+        }
+    }
+
+    public void queueMessage(String message) throws InterruptedException, StairwayExecutionException {
+        ByteString data = ByteString.copyFromUtf8(message);
+        PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(data).build();
+
+        try {
+            // Once published, returns a server-assigned message id (unique within the topic)
+            ApiFuture<String> future = publisher.publish(pubsubMessage);
+            String messageId = future.get();
+            logger.info("Queued message. Id: " + messageId + "; Msg: " + message);
+            } catch (ExecutionException ex) {
+            throw new StairwayExecutionException("Publish message failed", ex);
+        }
+    }
+
+    public void purgeQueue() {
+        PullRequest pullRequest =
+                PullRequest.newBuilder()
+                        .setMaxMessages(1)
+                        .setReturnImmediately(true)
+                        .setSubscription(subscriptionName)
+                        .build();
+
+        // Sometimes we get an empty response even when there are messages, so receive empty twice
+        // before calling it purged.
```

Comment:
Is this expected? and is this the recommended way to address the issue?

Reply:
The documentation is _very_ unclear on the semantics of the pull. So, no, this is not the recommended way to do it.

The purge is only used in tests to clean out the queue before running another test. I should document that. If we need "purge" in production, deleting the topic and re-creating it would be better.

Reply:
Now that I write that, what I _should_ do is make the subscription name and subscriber protected and move this to a subclass within the test directory.